cloud-checksum/src/io/copy/aws.rs

```diff
@@ -0,0 +1,244 @@
+//! AWS checksums and functionality.
+//!
+
+use crate::checksum::file::SumsFile;
+use crate::error::Error::ParseError;
+use crate::error::Result;
+use crate::io::copy::{CopyContent, ObjectCopy};
+use crate::io::Provider;
+use crate::MetadataCopy;
+use aws_sdk_s3::types::{MetadataDirective, TaggingDirective};
+use aws_sdk_s3::Client;
+use aws_smithy_runtime_api::client::orchestrator::HttpResponse;
+use aws_smithy_runtime_api::client::result::SdkError;
+use aws_smithy_types::byte_stream::ByteStream;
+use aws_smithy_types::error::metadata::ProvideErrorMetadata;
+use std::collections::HashMap;
+use tokio::io::AsyncReadExt;
+
+/// Build an S3 sums object.
+#[derive(Debug, Default)]
+pub struct S3Builder {
+    client: Option<Client>,
+    metadata_mode: MetadataCopy,
+}
+
+impl S3Builder {
+    /// Set the client.
+    pub fn with_client(mut self, client: Client) -> Self {
+        self.client = Some(client);
+        self
+    }
+
+    fn get_components(self) -> Result<(Client, MetadataCopy)> {
+        let error_fn = || {
+            ParseError(
+                "client, bucket, key and destinations are required in `S3Builder`".to_string(),
+            )
+        };
+
+        Ok((self.client.ok_or_else(error_fn)?, self.metadata_mode))
+    }
+
+    /// Set the copy metadata option.
+    pub fn with_copy_metadata(mut self, metadata_mode: MetadataCopy) -> Self {
+        self.metadata_mode = metadata_mode;
+        self
+    }
+
+    /// Build using the client, bucket and key.
+    pub fn build(self) -> Result<S3> {
+        Ok(self.get_components()?.into())
+    }
+}
+
+impl From<(Client, MetadataCopy)> for S3 {
+    fn from((client, metadata_mode): (Client, MetadataCopy)) -> Self {
+        Self::new(client, metadata_mode)
+    }
+}
+
+/// An S3 object and AWS-related existing sums.
+#[derive(Debug, Clone)]
+pub struct S3 {
+    client: Client,
+    metadata_mode: MetadataCopy,
+}
+
+impl S3 {
+    /// Create a new S3 object.
+    pub fn new(client: Client, metadata_mode: MetadataCopy) -> S3 {
+        Self {
+            client,
+            metadata_mode,
+        }
+    }
+
+    fn is_access_denied<T: ProvideErrorMetadata>(err: &SdkError<T, HttpResponse>) -> bool {
+        if let Some(err) = err.as_service_error() {
+            if err
+                .code()
+                .is_some_and(|code| code == "AccessDenied" || code == "InvalidSecurity")
+            {
+                return true;
+            }
+        }
+
+        false
+    }
```

Comment:
I'm not sure that this is the best way to handle this feature. There's three options - copy tags and fail, try copy tags but don't fail if permission is denied, and don't copy tags.

This will also ignore other permission errors, such as `PutObject` or `GetObject`, which might be unintuitive if the CLI succeeds but the copy failed. I think it might be better to just remove the "best-effort" option?

Otherwise, I think this CLI should do the copy, and then actually perform a "check" operation on the output to confirm that it is valid. 

I'm not sure if we want this in general or if it's better to leave this as a separate command. E.g. do we want to do `cloud-checksum copy <src> <dest>` or `cloud-checksum copy <src> <dest> && cloud-checksum check <src> <dest>` to confirm a valid copy?

Reply:
I'd keep it as deterministic as possible: copy tags vs no copy tags and report errors if either fails (in any object)... the best effort way is an unnecessary rabbit whole I reckon? :-S

Reply:
I think we have to do multi part first before deciding. Given some of the tag setting is going to be "extra" code - lets see how that plays out catering for different tag scenarios.

Maybe roman is right that the basic copy should just be deterministic - but there could be a flag ("--explicit-destination-check") that adds in a check after the copy